Libraries/LibCoreDump/Backtrace.cpp

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2020, Linus Groh <mail@linusgroh.de>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/LexicalPath.h>
+#include <AK/MappedFile.h>
+#include <AK/StringBuilder.h>
+#include <AK/Types.h>
+#include <LibCore/File.h>
+#include <LibCoreDump/Backtrace.h>
+#include <LibCoreDump/Reader.h>
+#include <LibELF/CoreDump.h>
+#include <LibELF/Image.h>
+
+namespace CoreDump {
+
+// FIXME: This cache has to be invalidated when libraries/programs are re-compiled.
+// We can store the last-modified timestamp of the elf files in ELFObjectInfo to invalidate cache entries.
+static HashMap<String, NonnullOwnPtr<ELFObjectInfo>> s_debug_info_cache;
+
+static const ELFObjectInfo* object_info_for_region(const ELF::Core::MemoryRegionInfo& region)
+{
+    auto name = region.object_name();
+
+    String path;
+    if (name.contains(".so"))
+        path = String::formatted("/usr/lib/{}", name);
+    else {
+        path = name;
+    }
+
+    if (auto it = s_debug_info_cache.find(path); it != s_debug_info_cache.end())
+        return it->value.ptr();
+
+    if (!Core::File::exists(path.characters()))
+        return nullptr;
+
+    MappedFile object_file(path);
+    if (!object_file.is_valid())
+        return nullptr;
+
+    auto info = make<ELFObjectInfo>(move(object_file), Debug::DebugInfo { make<ELF::Image>((const u8*)object_file.data(), object_file.size()) });
+    auto* info_ptr = info.ptr();
+    s_debug_info_cache.set(path, move(info));
+    return info_ptr;
```

Comment:
I assume the caller is allowed to use the pointer returned here? Is it okay to do a move and keep the pointer to the moved object around?

Reply:
Yes, it's a heap-allocated object, now owned by `s_debug_info_cache`.
We use the `info_ptr` temporary because `info` is nulled out by the `move()` into the hash map :)